internal/internal_worker.go

```diff
@@ -767,56 +763,42 @@ func (ae *activityExecutor) GetFunction() interface{} {
 
 func (ae *activityExecutor) Execute(ctx context.Context, input *commonpb.Payloads) (*commonpb.Payloads, error) {
 	fnType := reflect.TypeOf(ae.fn)
-	var args []reflect.Value
 	dataConverter := getDataConverterFromActivityCtx(ctx)
 
-	// activities optionally might not take context.
-	if fnType.NumIn() > 0 && isActivityContext(fnType.In(0)) {
-		args = append(args, reflect.ValueOf(ctx))
-	}
-
-	decoded, err := decodeArgs(dataConverter, fnType, input)
+	args, err := decodeArgsToRawValues(dataConverter, fnType, input)
 	if err != nil {
 		return nil, fmt.Errorf(
 			"unable to decode the activity function input payload with error: %w for function name: %v",
 			err, ae.name)
 	}
-	args = append(args, decoded...)
 
-	fnValue := reflect.ValueOf(ae.fn)
-	retValues := fnValue.Call(args)
-	return validateFunctionAndGetResults(ae.fn, retValues, dataConverter)
+	return ae.ExecuteWithActualArgs(ctx, args)
 }
 
-func (ae *activityExecutor) ExecuteWithActualArgs(ctx context.Context, actualArgs []interface{}) (*commonpb.Payloads, error) {
-	retValues := ae.executeWithActualArgsWithoutParseResult(ctx, actualArgs)
+func (ae *activityExecutor) ExecuteWithActualArgs(ctx context.Context, args []interface{}) (*commonpb.Payloads, error) {
 	dataConverter := getDataConverterFromActivityCtx(ctx)
 
-	return validateFunctionAndGetResults(ae.fn, retValues, dataConverter)
-}
-
-func (ae *activityExecutor) executeWithActualArgsWithoutParseResult(ctx context.Context, actualArgs []interface{}) []reflect.Value {
-	fnType := reflect.TypeOf(ae.fn)
-	var args []reflect.Value
-
-	// activities optionally might not take context.
-	argsOffeset := 0
-	if fnType.NumIn() > 0 && isActivityContext(fnType.In(0)) {
-		args = append(args, reflect.ValueOf(ctx))
-		argsOffeset = 1
-	}
-
-	for i, arg := range actualArgs {
-		if arg == nil {
-			args = append(args, reflect.New(fnType.In(i+argsOffeset)).Elem())
-		} else {
-			args = append(args, reflect.ValueOf(arg))
+	envInterceptor := getActivityEnvironmentInterceptor(ctx)
+	envInterceptor.fn = ae.fn
+
+	// Execute and serialize result
+	interceptor := envInterceptor.inboundInterceptor
+	if ae.skipInterceptors {
+		interceptor = envInterceptor
+	}
+	result, resultErr := interceptor.ExecuteActivity(ctx, &ExecuteActivityInput{Args: args})
+	var serializedResult *commonpb.Payloads
+	if result != nil {
+		// As a special case, if the result is already a payload, just use it
+		var ok bool
+		if serializedResult, ok = result.(*commonpb.Payloads); !ok {
+			var err error
+			if serializedResult, err = encodeArg(dataConverter, result); err != nil {
+				return nil, err
+			}
```

Comment:
Seems like this and the above bit (~ln 740) could still make use of a `serializeResults` helper function

Reply:
Unless I'm misunderstanding, it'd just be a couple line method to call `encodeArg`. I am not seeing the benefit of reuse. 

Also IIRC, to maintain functionality with existing code, I only allow activity result to bypass serialization if already the proto type, not workflow. Like many places in the code base, it is unfortunately inconsistent but the benefit of change is less than the potential cost of incompatibility.

Reply:
Oh, ok I thought that bypass might be fine in both places. Nevermind then.

Reply:
I think I added this "hack" when did a migration from `[]byte` to `Payloads`. It didn't make sense for me but w/o it something didn't work and I failed to track when this func is called two (or maybe more) times instead of 1. Basically, this is a protection against double encoding which needs to be fixed in the first place but I didn't have time for it.